scenarios/kubernetes_bazel.py

```diff
@@ -50,17 +54,93 @@ def get_version():
     with open('bazel-genfiles/version') as fp:
         return fp.read().strip()
 
+def get_changed(base, pull):
+    """Get affected packages between base sha and pull sha."""
+    diff = check_output(
+        'git', 'diff', '--name-only',
+        '--diff-filter=d', '%s...%s' % (base, pull))
+    return check_output(
+        'bazel', 'query',
+        '--noshow_progress',
+        'set(%s)' % diff).split('\n')
+
+def query(kind, selected_pkgs, changed_pkgs):
+    """
+    Run a bazel query against target kind, include targets from args.
+
+    Returns a list of kind objects from bazel query.
+    """
+
+    # Changes are calculated and no packages found, return empty list.
+    if changed_pkgs == []:
+        return []
+
+    selection = '//...'
+    if selected_pkgs:
+        selection = 'set(%s)' % ' '.join(selected_pkgs)
+
+    changes = '//...'
+    if changed_pkgs:
+        changes = 'set(%s)' % ' '.join(changed_pkgs)
+
+    return check_output(
+        'bazel', 'query',
+        '--keep_going',
+        '--noshow_progress',
+        'kind(%s, rdeps(%s, %s))' % (kind, selection, changes)
+    ).split('\n')
+
+def clean_file_in_dir(dirname, filename):
+    """Recursively remove all file with filename in dirname."""
+    for parent, _, filenames in os.walk(dirname):
+        for name in filenames:
+            if name == filename:
+                os.remove(os.path.join(parent, name))
+
 def main(args):
     """Trigger a bazel build/test run, and upload results."""
+    # pylint:disable=too-many-branches, too-many-statements, too-many-locals
+    if args.install:
+        for install in args.install:
+            if not os.path.isfile(install):
+                raise ValueError('Invalid install path: %s' % install)
+            check('pip', 'install', '-r', install)
+
     check('bazel', 'clean', '--expunge')
     res = 0
     try:
+        affected = None
+        if args.affected:
+            base = os.getenv('PULL_BASE_SHA', '')
+            pull = os.getenv('PULL_PULL_SHA', '')
+            if not base or not pull:
+                raise ValueError('PULL_BASE_SHA and PULL_PULL_SHA must be set!')
+            affected = get_changed(base, pull)
+
+        build_pkgs = None
+        test_pkgs = None
         if args.build:
-            check('bazel', 'build', *args.build.split(' '))
+            build_pkgs = args.build.split(' ')
+        if args.test:
+            test_pkgs = args.test.split(' ')
+
+        if build_pkgs or affected:
+            buildables = query('.*_binary', build_pkgs, affected)
+            if buildables:
+                check('bazel', 'build', *buildables)
+
         if args.release:
             check('bazel', 'build', *args.release.split(' '))
-        if args.test:
-            check('bazel', 'test', *args.test.split(' '))
+
+        # clean up previous test.xml
+        clean_file_in_dir('./bazel-testlogs', 'test.xml')
```

Comment:
also, I think we may need to guarantee that bazel has run before we get here, as otherwise this symlink may not exist.

in the shell script we did this with a bazel build with no arguments. maybe fold it into L130 and run bazel build always, regardless of whether there are buildables or whether --build was set?